openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -63,43 +73,72 @@ def get_next_query_params(self, prev_query_params=None, **kwargs):
             next_offset = prev_query_params["offset_num"] + self.batch_limit
             return {"offset_num": next_offset}
 
-    def get_response_json(self, query_params: dict):
-        """
-        Call the SQL to pull json from Postgres, where the raw data has been loaded.
-        """
-        sql_string = self.sql_template.format(
-            batch_limit=self.batch_limit, offset_num=query_params["offset_num"]
-        )
-        sql_result = self.pg.get_records(sql_string)
-        # Postgres returns a list of tuples, even if it's one tuple with one item.
-        return sql_result[0][0]
+    def get_response_json(self, query_params):
+        raise NotImplementedError("TSV files from AWS S3 processed in postgres.")
 
     def get_batch_data(self, response_json):
-        if response_json:
-            return response_json
-        return None
+        raise NotImplementedError("TSV files from AWS S3 processed in postgres.")
 
     def get_record_data(self, data):
-        if data.get("foreign_identifier") is None:
-            # TO DO: maybe raise an error here or after a certain number of these?
-            # more info in https://github.com/WordPress/openverse-catalog/issues/684
-            return None
-        license_url = data.get("license_url")
-        license_info = get_license_info(license_url=license_url)
-        if license_info == NO_LICENSE_FOUND:
-            return None
-        record_data = {k: data[k] for k in data.keys() if k != "license_url"}
-        record_data["license_info"] = license_info
-        return record_data
+        raise NotImplementedError("TSV files from AWS S3 processed in postgres.")
 
     def get_media_type(self, record):
         # This provider only supports Images via S3, though they have some audio files
         # on site and in the API.
-        return "image"
+        # I'm using a constant here, which maybe should just be a class constant, but
+        # it's there to work with the static methods.
+        return MEDIA_TYPE
 
     def endpoint(self):
         raise NotImplementedError("Normalized TSV files from AWS S3 means no endpoint.")
 
+    @staticmethod
+    def load_transformed_data(
+        intermediate_table: str,
+        page_length: int,
+        postgres_conn_id=POSTGRES_CONN_ID,
+        sql_template_file_name="transformed_table.template.sql",
+        identifier: str = None,
+    ):
+        pg = PostgresHook(postgres_conn_id)
+        sql_template = (SCRIPT_DIR / sql_template_file_name).read_text()
+        total_records = 0
+        # get the number of "pages" in the photos table
+        max_id = pg.get_records("SELECT max(photo_id) FROM inaturalist.photos")[0][0]
+        total_pages = int(max_id / page_length) + 1
+        logger.info(f"Loading {intermediate_table}...")
+        logger.info(f"Max photo_id={max_id}. Transform {page_length} ids / iteration.")
+        # then cycle through the pages
+        for page_start in range(0, max_id, page_length):
```

Comment:
I haven't used this myself yet so this idea would need exploration, but I wonder if this would be a good place to try using [dynamic tasks](https://docs.astronomer.io/learn/dynamic-tasks). Using task mapping, I think we could:
* have a task to get the number of "pages"
* expand a `loader` task to create a separate task instance for each "page"

That would increase the opportunity for parallelization (maybe, depending on the handling of intermediate tables 🤔), but even without concurrency it would be nice to let us clear/retry individual page tasks rather than having one giant task.

Reply:
Right. I guess before I dig in too far, how much do we want to get into parallelism outside of pg basics? I guess I feel like we stopped at ~15 million records in 5 days or something like that with the current code, and this step gets us to ~120 million records in 2 days, maybe we could just call it a win for now, and circle back if needed? I'm definitely curious about dynamically generating tasks, but I'm really trying to avoid space constraints, so right now the one temp table is really treated as a temporary pass through space and emptied out after each page is upserted, which definitely will not work if we're trying to run pages in parallel. Just inaturalist is ~15% of the existing `image` table size and I wasn't able to make holding two copies work locally. And creating separate temp tables per page feels like it would be a big lift in terms of the current code, and maybe not worth it? Though I do see the potential for real reporting and operational wins of separate tasks even if they are run sequentially. ???

Reply:
That's some helpful perspective Rebecca, thanks! I think with that in mind the current changeset is worth merging as it stands, and perhaps that other parallelization can be explored down the road.

Reply:
Definitely agree about parallelization being a v2 concern (_if_ it is necessary at all -- the estimated times you've given are great and it's not at all clear to me that it's worth it yet! 🤩), I mean to suggest that the _opportunity_ for easier future parallelization was one (small) advantage of splitting this up. The much greater immediate advantage in my mind is the ability to retry/inspect individual tasks. I don't think that would require dealing with any of the tricky space constraints or handling parallel load tables.

As it stands, this is such a phenomenal improvement that I do think it would be worth merging on its own. As we move to splitting the typical ingestion flow into smaller tasks (splitting the load and cleaning steps for example), and as we've been dealing with a lot of long-running DAG failures, I'm just slightly cautious about very large tasks.

Reply:
Yeah, I 100% hear you on the opportunity to look at the different tasks, @stacimc. I tried to address that need with some logging, but it's clunky, no question.

I just started to try coding this, and it occurred to me that for it to work, I would need a separate dag for the preingestion tasks rather than a task group within `inaturalist_workflow`. The number of pages (and therefore the number of tasks we would want to see in the dag) is based on the `max(photo_id)` loaded to `inaturalist.photos`, but with a single dag, that load would not have happened yet when airflow puts the dag task list together.

That said, I think that it's probably pretty common practice to separate out staging and target loads into separate dags, and we could have something like `inaturalist_staging_workflow` and `inaturalist_target_workflow`. But I wonder if that could be a v2 improvement?

Reply:
Separate DAGs would definitely be a heavy lift, but I think task mapping by itself should accomplish what we want here. Task mapping allows you to generate the task list totally dynamically, so you _don't_ need to know the number of tasks you want to see in the DAG in advance. The [documentation](https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html#task-generated-mapping) even mentions building the list of parameters from API calls or DB queries. There's really good instructions [here](https://docs.astronomer.io/learn/dynamic-tasks#map-inputs-when-both-tasks-are-defined-with-traditional-operators).

I wouldn't say this is a blocker, but it would be nice to have if it's not too much lift. I played around with this just to make sure I understood and I was able to get a rough version working. Basically you just have a function that gets the list of pages:

```
    @staticmethod
    def get_pages(
        intermediate_table: str,
        page_length: int,
        postgres_conn_id=POSTGRES_CONN_ID,
    ):
        pg = PostgresHook('postgres_conn_id')
        # get the number of "pages" in the photos table
        max_id = pg.get_records("SELECT max(photo_id) FROM inaturalist.photos")[0][0]
        total_pages = int(max_id / page_length) + 1

        # Return the list of pages. Note the weird formatting because this is being passed to op_args, which expects each arg to be a list
        return [[x] for x in range(0, max_id, page_length)]
```

And then a function that takes in `page_start` and just does the processing for a single page:

```
    @staticmethod
    def load_transformed_data(
        page_start,
        intermediate_table: str = '',
        page_length: int = 0,
        postgres_conn_id=POSTGRES_CONN_ID,
        sql_template_file_name="transformed_table.template.sql",
        identifier: str = None,
    ):

        pg = PostgresHook(postgres_conn_id)
        sql_template = (SCRIPT_DIR / sql_template_file_name).read_text()
        total_records = 0
        page_end = page_start + page_length - 1
        page_log = f"page {int(page_start / page_length) + 1}"
        logger.info(f"Starting at photo_id {page_start}, {page_log}")
        (transformed_records, max_id_loaded) = pg.get_records(
...and so on, truncating the table at the end
```

And then when you define the tasks, you use `expand` to dynamically map a bunch of `load_transformed_data` tasks using the result of the `get_pages` task:

```
                from airflow import XComArg

                get_pages = PythonOperator(
                    task_id="get_pages",
                    python_callable=INaturalistDataIngester.get_pages,
                    op_kwargs={ <omitted for brevity> }
                )

                load_transformed_data = PythonOperator.partial(
                    task_id="load_transformed_data",
                    python_callable=INaturalistDataIngester.load_transformed_data,
                    op_kwargs={
                        "intermediate_table": sql._get_load_table_name(
                            LOADER_ARGS["identifier"]
                        ),
                        "page_length": 2_000_000,
                        "postgres_conn_id": LOADER_ARGS["postgres_conn_id"],
                        "identifier": LOADER_ARGS["identifier"],
                    },
                ).expand(op_args=XComArg(get_pages))  # This part expands it using the return_value of get_pages
```

And then in the UI it looked like this for me:

<img width="663" alt="Screen Shot 2022-12-21 at 1 24 25 PM" src="https://user-images.githubusercontent.com/63313398/209005194-05779187-ad21-4800-a35d-7c8757847d2f.png">

What do you think? I'll defer to you here 😄 We can certainly iterate.

Reply:
Oh wow, I swear I at least skimmed that article the first time you linked to it, but I totally missed the xcom args part. Yes, this is a million times better. Let's do this. I'll get right on it.

Reply:
OK, and just to confirm, two follow-up questions:

- we'll use `max_active_tis_per_dag` to make sure that only one of these things is running at a time (see https://docs.astronomer.io/learn/dynamic-tasks), so they don't truncate each other, and it's ok to set it for all dags? and 
- this only applies to tasks, not task groups, so right it still needs to have all the load --> clean --> upsert --> truncate steps within that one function.

True?

Reply:
> we'll use max_active_tis_per_dag to make sure that only one of these things is running at a time

Yep! You can set this _on the actual task_, as in passing it in to the `PythonOperator` for the `load_transformed_data` task definition [Example](https://docs.astronomer.io/learn/airflow-scaling-workers#task-level-airflow-settings). So that way you're only configuring the max active task instances of this _particular_ task, per DagRun.

> this only applies to tasks, not task groups, so right it still needs to have all the load --> clean --> upsert --> truncate steps within that one function.

I think you can actually generate [whole task groups dynamically](https://docs.astronomer.io/learn/task-groups#generate-task-groups-dynamically-at-runtime)... but I think you need to be using the TaskFlow api. I suspect the UI will also get pretty confusing. So I think it makes total sense to leave this as one function for now!

You might've seen this already, but I also found this discussion about [consolidating XComs from dynamic tasks](https://forum.astronomer.io/t/dynamic-task-mapping-xcom-handling/2003/2) which might be helpful for summing the counts from each page.